src/main/java/jenkins/plugins/logstash/persistence/LogzioDao.java

```diff
@@ -0,0 +1,127 @@
+
+import java.io.*;
+import java.util.*;
+import java.util.concurrent.Executors;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import jenkins.model.Jenkins;
+import jenkins.plugins.logstash.LogstashConfiguration;
+import net.sf.json.JSONArray;
+import net.sf.json.JSONObject;
+
+/**
+ * Logz.io Data Access Object.
+ *
+ * @author Ido Halevi
+ */
+
+public class LogzioDao extends AbstractLogstashIndexerDao {
+    private static final int CONNECT_TIMEOUT = 10*1000;
+    private static final int CORE_POOL_SIZE = 2;
+    private static final int DRAIN_TIMEOUT = 2;
+    private static final int FS_PERCENT_THRESHOLD = 99;
+    private static final int GC_PERSISTED_QUEUE_FILE_INTERVAL_SECOND = 30;
+    private static final int SOCKET_TIMEOUT = 10*1000;
+    private static final String TYPE = "jenkins_plugin";
+    private final LogzioSender logzioSender;
+
+    private String key;
+    private String host;
+
+    //primary constructor used by indexer factory
+    public LogzioDao(String host, String key){
+        this(null, host, key);
+    }
+
+    // Factored for unit testing
+    LogzioDao(LogzioSender factory, String host, String key){
+        this.host = host;
+        this.key = key;
+
+        // create file for sender queue
+        File fp;
+        Jenkins jenkins = Jenkins.getInstanceOrNull();
+        if (jenkins == null){
+            // for testing (mvn package)
+            fp = Files.createTempDir();
+        }else{
+            fp = new File(jenkins.getRootDir() + "/logs/logzio_plugin");
+            boolean folderExisted = fp.mkdirs() || fp.setWritable(true);
+            if (!folderExisted) {
+                throw new IllegalArgumentException("Unable to create path");
+            }
+        }
+
+        try{
+            this.logzioSender = factory == null ? LogzioSender.getOrCreateSenderByType(key, TYPE, DRAIN_TIMEOUT,
+                    FS_PERCENT_THRESHOLD, fp, host, SOCKET_TIMEOUT, CONNECT_TIMEOUT,false, null,
+                    Executors.newScheduledThreadPool(CORE_POOL_SIZE),GC_PERSISTED_QUEUE_FILE_INTERVAL_SECOND, true) : factory;
+            this.logzioSender.start();
+        }catch (LogzioParameterErrorException e){
+            throw new IllegalArgumentException(e.getMessage());
+        }
+    }
+
+    @Override
+    public void push(String data) {
+        JSONObject jsonData = JSONObject.fromObject(data);
+        JSONArray logMessages = jsonData.getJSONArray("message");
+        for (Object logMsg : logMessages) {
+            JsonObject logLine = createLogLine(jsonData, logMsg.toString());
+            this.logzioSender.send(logLine);
+        }
+    }
+
+    protected JsonObject createLogLine(JSONObject jsonData, String logMsg) {
+        JsonObject logLine = new JsonObject();
+        logLine.addProperty("message", logMsg);
+        logLine.addProperty("@timestamp", LogstashConfiguration.getInstance().getDateFormatter().format(Calendar.getInstance().getTime()));
```

Comment:
the timestamps will point to the time the log lines were processed here, not to the time they were emmited

also: why use Calendar instead of just System.currentTimeMilis?

Reply:
The problem I'm trying to solve is one that will show up in ES as well. The plugin gives all the logs the same timestamp and this creates a situation where the logs are not in order in Kiban. To solve it we need a field '@timestamp' in each log that is different from the other.